pkg/storage/db.go

```diff
@@ -199,6 +208,79 @@ func (db *DB) ping() error {
 	return nil
 }
 
+// GeneratePartition finds the latest partition, and generate the next partition based on the previous partions time range
+func (db *DB) GeneratePartition(ctx context.Context) error {
+	stmt := "SELECT partition_end FROM partitions ORDER BY partition_begin DESC LIMIT 1"
+	row := db.sqldb.QueryRowContext(ctx, stmt)
```

Comment:
Suggestion: There are a number of conditions that could result in concurrent executions of this path even though we're relying on an explicit trigger via a special endpoint. I'd recommend looking at some of the locking options (https://www.postgresql.org/docs/11/sql-lock.html) and putting all of the partition management work inside a transaction. This would enable you to bail early if unable to acquire the lock and let the other running call finish.